### PR TITLE
feat(onboarding): use cache to check if repo is onboarded

### PR DESCRIPTION
**Pull request by RahulGautamSingh**
## Changes
- Added `onboardingBranchCache` that contains `defaultBranchSha` and `onboardingBranchSha`
- If both SHAs are same as last run repo isn't onboarded
- If anyone is different we fall back to the normal `isOnboarded` computation logic ie. cloning

## Context
- Part of: https://github.com/renovatebot/renovate/issues/17839

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

**Comment by RahulGautamSingh** [comment]
Pls don''t review now. Some work is pending.

**Comment by RahulGautamSingh** [comment]
Blockedd by https://github.com/renovatebot/renovate/pull/20746

**Comment by rarkins** [comment]
So this satisfies just the use case where there's no commits to the default branch or the onboarding branch since the last time we ran?

**Comment by RahulGautamSingh** [comment]
Yes

File: lib/workers/repository/onboarding/branch/check.ts [hunk 1]
```diff
@@ -61,7 +61,21 @@ export async function isOnboarded(config: RenovateConfig): Promise<boolean> {
...
     return true;
   }
+
   const cache = getCache();
+  const onboardingBranchCache = cache?.onboardingBranchCache;
+  // if onboarding cache is present and base branch has not been updated branch is not onboarded
+  if (
+    onboardingBranchCache &&
+    onboardingBranchCache.defaultBranchSha ===
+      getBranchCommit(config.defaultBranch!)! &&
```
Comment (viceice) [inline lib/workers/repository/onboarding/branch/check.ts:71]:
```suggestion
      getBranchCommit(config.defaultBranch!) &&
```
i don't think we need it for compare

File: lib/workers/repository/onboarding/branch/onboarding-branch-cache.spec.ts [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+  });
+
+  it('updates old cache', () => {
+    const dummyCache = {
+      onboardingBranchCache: {
+        branchName: 'configure/renovate',
+        defaultBranchSha: 'default-sha',
+        onboardingBranchSha: 'onboarding-sha',
+      },
+    } as RepoCacheData;
```
Comment (viceice) [inline lib/workers/repository/onboarding/branch/onboarding-branch-cache.spec.ts:33]:
```suggestion
    } satisfies RepoCacheData;
```

File: lib/workers/repository/onboarding/branch/onboarding-branch-cache.spec.ts [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+  deleteOnboardingCache,
+  setOnboardingCache,
+} from './onboarding-branch-cache';
+
+jest.mock('../../../../util/cache/repository');
+const cache = mocked(_cache);
+
+describe('workers/repository/onboarding/branch/onboarding-branch-cache', () => {
+  it('sets new cache', () => {
+    const dummyCache = {} as RepoCacheData;
```
Comment (viceice) [inline lib/workers/repository/onboarding/branch/onboarding-branch-cache.spec.ts:14]:
```suggestion
    const dummyCache = {} satisfies RepoCacheData;
```

File: lib/workers/repository/onboarding/branch/onboarding-branch-cache.spec.ts [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+  });
+
+  it('deletes cache', () => {
+    const dummyCache = {
+      onboardingBranchCache: {
+        branchName: 'configure/renovate',
+        defaultBranchSha: 'default-sha',
+        onboardingBranchSha: 'onboarding-sha',
+      },
+    } as RepoCacheData;
```
Comment (viceice) [inline lib/workers/repository/onboarding/branch/onboarding-branch-cache.spec.ts:56]:
```suggestion
    } satisfies RepoCacheData;
```

File: lib/workers/repository/onboarding/branch/index.ts [hunk 5]
```diff
@@ -78,6 +96,13 @@ export async function checkOnboardingBranch(
         { branch: onboardingBranch, commit, onboarding: true },
         'Branch created'
       );
+
+      // set onboarding branch cache
+      setOnboardingCache(
+        config.onboardingBranch!,
+        getBranchCommit(config.defaultBranch!)!,
+        getBranchCommit(config.onboardingBranch!)!
```
Comment (viceice) [inline lib/workers/repository/onboarding/branch/index.ts:104]: don't we already have the Sha above?

File: lib/workers/repository/onboarding/branch/index.ts [hunk 6]
```diff
@@ -47,6 +58,13 @@ export async function checkOnboardingBranch(
         { branch: config.onboardingBranch, commit, onboarding: true },
         'Branch updated'
       );
+
+      // update onboarding cache
+      setOnboardingCache(
+        config.onboardingBranch!,
+        getBranchCommit(config.defaultBranch!)!,
+        getBranchCommit(config.onboardingBranch!)!
```
Comment (viceice) [inline lib/workers/repository/onboarding/branch/index.ts:66]: don't we have the commit Sha in a variable above?

**Comment by RahulGautamSingh** [comment]
Will be needing more logging For when onboarding cache deleted, created & updated? 

During test repo runs it was hard for me to see logs and say if the cache is working perfectly.
The only log statement I was using is this 
```logger.debug('Onboarding cache is valid. Repo is not onboarded');```

File: lib/util/cache/repository/types.ts [hunk 7]
```diff
@@ -26,6 +26,12 @@ export interface BranchUpgradeCache {
   sourceUrl?: string;
 }
 
+export interface OnboardingBranchCache {
+  branchName: string;
```
Comment (rarkins) [inline lib/util/cache/repository/types.ts:30]:
```suggestion
  onboardingBranch: string;
```

Comment (rarkins) [reply lib/util/cache/repository/types.ts:30]: Are you using it/ checking it later once it's set?

**Review by rarkins — CHANGES_REQUESTED** [review]

File: lib/util/cache/repository/types.ts (hunk 7 above)
Comment (RahulGautamSingh) [reply lib/util/cache/repository/types.ts:30]: The branchName? No.

Comment (rarkins) [reply lib/util/cache/repository/types.ts:30]: So should it be left out?

**Review by viceice — APPROVED** [review]
code LGTM

File: lib/util/cache/repository/types.ts (hunk 7 above)
Comment (RahulGautamSingh) [reply lib/util/cache/repository/types.ts]: Can be.

**Comment by RahulGautamSingh** [comment]
@rarkins I think this can be merged before https://github.com/renovatebot/renovate/pull/20893
Integrating this cache simple cache-implementation in the linked PR won't be complicated as I had previously assumed.

**Review by rarkins — APPROVED** [review]

**Pull request merged**